rustls/src/time_provider.rs

```diff
@@ -0,0 +1,29 @@
+//! no-std time provider
+
+use core::fmt::Debug;
+
+use pki_types::UnixTime;
+
+/// An object that provides the current time.
+///
+/// This is used to, for example, check if a certificate has expired during certificate validation,
+/// or to check the age of a ticket.
+#[allow(unreachable_pub)]
+pub trait TimeProvider: Debug + Send + Sync {
+    /// Returns the current time
+    ///
+    /// Or `None` if unable to retrieve it
+    fn current_time(&self) -> Option<UnixTime>;
+}
+
+#[derive(Debug)]
+pub(crate) struct DefaultTimeProvider;
+
+impl TimeProvider for DefaultTimeProvider {
+    fn current_time(&self) -> Option<UnixTime> {
+        #[cfg(feature = "std")]
+        return Some(UnixTime::now());
+        #[cfg(not(feature = "std"))]
```

Comment:
As explained [previously](https://github.com/rustls/rustls/pull/1502#discussion_r1434942346), I'm not a fan of this setup. Can we instead feature-gate the `DefaultTimeProvider` on `#[cfg(feature = "std")]`, forcing `no_std` users to supply their own?

Reply:
Also, if every caller of `current_time()` is going to convert the `Option` to `Result` anyway, should we just allow the implementer of this trait to supply an `Error` directly?

Reply:
> Can we instead feature-gate the DefaultTimeProvider on #[cfg(feature = "std")], forcing no_std users to supply their own?

I can give this a shot. @ctz Do you prefer this approach as well?

> I suppose the no_std users who are the most likely implementers don't currently have a Box<dyn std::error::Error> to smuggle an alternative error through...

Yeah, that seems like a good reason to leave it as an `Option` IMO.